### PR TITLE
Ethereum QML types: #612

### PR DESCRIPTION
**Pull request by yann300**
- Display an Ether value (including units and formatting).
- Manage big integer in QML side.

### CODE REVIEW

File: mix/ClientModel.h [hunk 1]
```diff
@@ -26,6 +26,7 @@
 #include <atomic>
 #include "DebuggingStateWrapper.h"
 #include "MixClient.h"
+#include "QEther.h"
```
Comment (arkpar) [inline mix/ClientModel.h]: Is it needed here? Maybe a forward declaration will do?

File: mix/DebuggingStateWrapper.cpp [hunk 2]
```diff
@@ -66,25 +67,19 @@ std::tuple<QList<QObject*>, QQMLMap*> DebuggingStateWrapper::getHumanReadableCod
 	return std::make_tuple(codeStr, QPointer<QQMLMap>(new QQMLMap(codeMapping)));
 }
 
-QString DebuggingStateWrapper::gasCost()
+QBigInt* DebuggingStateWrapper::gasCost()
 {
-	std::ostringstream ss;
-	ss << std::dec << m_state.gasCost;
-	return QString::fromStdString(ss.str());
+	return new QBigInt(m_state.gasCost, this);
```
Comment (arkpar) [inline mix/DebuggingStateWrapper.cpp]: You should set ownership to javascript to avoid memory leaks

File: mix/QBigInt.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/** @file QBigInt.cpp
+ * @author Yann yann@ethdev.com
+ * @date 2015
+ */
+
+#include "boost/variant/multivisitors.hpp"
```
Comment (arkpar) [inline mix/QBigInt.cpp]: angle brackets?

File: mix/ClientModel.h (hunk 1 above)
Comment (yann300) [inline mix/ClientModel.h]: OK I moved this include in the right header file.

File: mix/QBigInt.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	Q_INVOKABLE QBigInt* subtract(QBigInt* const& _value) const;
+	/// Add @a _value to the current big integer. Invokable from QML.
+	Q_INVOKABLE QBigInt* add(QBigInt* const& _value) const;
+	/// Multiply by @a _value. Invokable from QML.
+	Q_INVOKABLE QBigInt* multiply(QBigInt* const& _value) const;
+	/// divide by @a _value. Invokable from QML.
+	Q_INVOKABLE QBigInt* divide(QBigInt* const& _value) const;
+
+protected:
+	 BigIntVariant m_internalValue;
```
Comment (arkpar) [inline mix/QBigInt.h]: Using variant here makes code quite complex. Maybe use template base class instead?

Comment (yann300) [inline mix/QBigInt.h]: Yes I tried but we cannot use Template with a QObject class. And with
Template, we have to know the type to use (bigint or u256) at compile time.
With boost::variant. We just have to set a pool of type to manage and the
rest is done during the runtime.

2015-01-19 16:48 GMT+01:00 Arkadiy Paronyan <notifications@github.com>:

**Pull request merged**